libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxSpeechLoader.kt

```diff
@@ -17,28 +20,92 @@ import com.mapbox.navigation.ui.utils.internal.resource.ResourceLoader
-    private val resourceLoader: ResourceLoader
+    private val resourceLoader: ResourceLoader,
 ) {
 
-    suspend fun load(typeAndAnnouncement: TypeAndAnnouncement): Expected<Throwable, ByteArray> {
+    private val currentRequests = mutableSetOf<Long>()
+    private val downloadedInstructions = mutableSetOf<TypeAndAnnouncement>()
+    private val downloadedInstructionsLock = Any()
+    private val defaultScope = InternalJobControlFactory.createDefaultScopeJobControl().scope
+
+    suspend fun load(
+        voiceInstruction: VoiceInstructions,
+        onlyCache: Boolean
+    ): Expected<Throwable, ByteArray> {
         return runCatching {
-            val url = instructionUrl(typeAndAnnouncement.announcement, typeAndAnnouncement.type)
-            val response = resourceLoader.load(url)
+            val typeAndAnnouncement = VoiceInstructionsParser.parse(voiceInstruction)
+                .getValueOrElse { throw it }
+            val request = createRequest(typeAndAnnouncement).apply {
+                if (onlyCache) {
+                    networkRestriction = NetworkRestriction.DISALLOW_ALL
+                }
+            }
+            val response = resourceLoader.load(request)
             return processResponse(response)
         }.getOrElse {
             createError(it)
         }
     }
 
-    private suspend fun ResourceLoader.load(url: String) = load(ResourceLoadRequest(url))
+    fun triggerDownload(voiceInstructions: List<VoiceInstructions>) {
+        defaultScope.launch {
+            voiceInstructions.forEach { voiceInstruction ->
+                val typeAndAnnouncement = VoiceInstructionsParser.parse(voiceInstruction).value
+                if (typeAndAnnouncement != null && !hasTypeAndAnnouncement(typeAndAnnouncement)) {
+                    predownload(typeAndAnnouncement)
+                }
+            }
+        }
+    }
+
+    fun cancel() {
+        defaultScope.cancel()
+        currentRequests.forEach { resourceLoader.cancel(it) }
+    }
+
+    private fun hasTypeAndAnnouncement(typeAndAnnouncement: TypeAndAnnouncement): Boolean {
+        synchronized(downloadedInstructionsLock) {
+            return typeAndAnnouncement in downloadedInstructions
+        }
+    }
+
+    private fun predownload(typeAndAnnouncement: TypeAndAnnouncement) {
+        try {
+            val request = createRequest(typeAndAnnouncement)
+            var id: Long? = null
+            id = resourceLoader.load(request) { result ->
+                id?.let { currentRequests.remove(it) }
+                // tilestore thread
+                if (result.isValue) {
+                    synchronized(downloadedInstructionsLock) {
+                        downloadedInstructions.add(typeAndAnnouncement)
+                    }
+                }
+            }
+            currentRequests.add(id)
+        } catch (ex: Throwable) {
+            logE("Failed to download instruction '$typeAndAnnouncement': ${ex.localizedMessage}")
+        }
+    }
```

Comment:
Continuation of: https://github.com/mapbox/mapbox-navigation-android/pull/6547#discussion_r1053099007

> But then the requests won't be run in parallel?

They will be in parallel if you run new coroutine per voice instruction:

```
launch {
    predownload(typeAndAnnouncement: TypeAndAnnouncement)
}
```

Reply:
Oh, I see what you mean. Yes, I don't need `currentRequests` then, thanks.